File: guest/arch/kernel-upgrade/kernel_upgrade.py
#!/usr/bin/env python3
"""Upgrade only the Arch AVF kernel, preserving its disk and a verified rollback copy."""
import argparse
import hashlib
import importlib.util
import json
from pathlib import Path
import re
import shlex
import shutil
import subprocess
import uuid


def verify_artifact(root, commit):
    if not re.fullmatch(r"[0-9a-f]{40}", commit):
        raise ValueError("Use the full, independently selected CI source commit")
    # Reuse the exact probe manifest/driver checks that validated this kernel.
    probe = Path(__file__).resolve().parent.parent / "virtiofs-probe"
    import sys
    sys.path.insert(0, str(probe))
    spec = importlib.util.spec_from_file_location("probe_artifact", probe / "run_device.py")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    expected = module.verify_artifact(root)
    provenance = (root / "probe-provenance.txt").read_text()
    if provenance != f"commit={commit}\npurpose=diskless-virtiofs-probe\n":
        raise ValueError("Artifact source commit/purpose does not match")
    return expected["Image"]


def require_stopped(status, *, rollback=False):
    states = ("stopped", "error") if rollback else ("stopped",)
    if (status.get("status") not in states or status.get("running") is not False
            or status.get("active_sessions") != 0
            or (not rollback and status.get("clean_shutdown") is not True)):
        raise RuntimeError("Arch must be stopped with no sessions (new installs also require clean shutdown); "
                           "use termux-arch-vm --stop first")


def status():
    result = subprocess.run(["termux-arch-vm", "--status"], check=True, capture_output=True,
                            text=True, timeout=30)
    return json.loads(result.stdout)


def shell(command):
    subprocess.run(["rish", "-c", command], check=True, timeout=90)


def sha(path):
    with path.open("rb") as source:
        return hashlib.file_digest(source, "sha256").hexdigest()


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--helper", type=Path, required=True, help="Trusted kernel-upgrade.jar built from this source")
    sub = parser.add_subparsers(dest="operation", required=True)
    sub.add_parser("inspect", help="Inspect the installed kernel while Arch is stopped")
    install = sub.add_parser("install", help="Install a verified, extracted Virtio-FS CI kernel")
    install.add_argument("artifact", type=Path)
    install.add_argument("--commit", required=True, help="Exact source commit of the successful CI run")
    rollback = sub.add_parser("rollback", help="Restore a saved kernel; retain the current one too")
    rollback.add_argument("backup", help="32-character backup ID printed by install")
    args = parser.parse_args()
    if not args.helper.is_file() or args.helper.is_symlink():
        parser.error("Helper must be a trusted regular jar file")
    candidate = verify_artifact(args.artifact, args.commit) if args.operation == "install" else "-"
    if args.operation == "rollback" and not re.fullmatch(r"[0-9a-f]{32}", args.backup):
        parser.error("Invalid backup ID")
    if args.operation != "inspect":
        require_stopped(status(), rollback=args.operation == "rollback")
    token = uuid.uuid4().hex
    public = Path("/storage/emulated/0/Download") / f"aether-kernel-{token}"
    stage = f"/data/local/tmp/aether-kernel-{token}"
    state = Path.home() / ".local/state/termux-aether" / f"kernel-{token}"
    state.mkdir(parents=True, mode=0o700)
    public.mkdir()
    report = {"status": "failed", "operation": args.operation, "state": str(state),
              "stage": stage, "public": str(public)}
    success = False
    try:
        inputs = {"kernel-upgrade.jar": args.helper}
        if args.operation == "install":
            inputs["Image"] = args.artifact / "Image"
        expected = {name: sha(path) for name, path in inputs.items()}
        if args.operation == "install" and expected["Image"] != candidate:
            raise ValueError("Artifact changed after verification")
        for name, path in inputs.items():
            shutil.copyfile(path, public / name)
        copies = " && ".join(f"cp {shlex.quote(str(public / name))} {stage}/{name}" for name in inputs)
        checksums = "".join(f"{digest}  {name}\n" for name, digest in expected.items())
        check = f"printf '%s' {shlex.quote(checksums)} | sha256sum -c - > integrity.txt"
        shell(f"umask 077 && mkdir {stage} && {copies} && cd {stage} && {check} && chmod 0444 kernel-upgrade.jar")

        def invoke(operation, current="-", wanted="-", source="-"):
            arguments = " ".join(shlex.quote(value) for value in (operation, stage, current, wanted, source))
            # Complete results travel as files: rish stdout can lose chunks on this firmware.
            shell(f"cd {stage} && rm -f result.json && {check} && "
                  f"CLASSPATH={stage}/kernel-upgrade.jar app_process /system/bin KernelUpgrade {arguments} "
                  f"> {stage}/{operation}.log 2>&1; upgrade_rc=$?; "
                  f"printf '%s\\n' \"$upgrade_rc\" > {stage}/exit.txt; "
                  f"cp {stage}/exit.txt {shlex.quote(str(public))}/exit.txt; "
                  f"cp {stage}/{operation}.log {shlex.quote(str(public))}/{operation}.log; "
                  f"if [ -f {stage}/result.json ]; then cp {stage}/result.json {shlex.quote(str(public))}/result.json; fi")
            shutil.copyfile(public / f"{operation}.log", state / f"{operation}.log")
            if (public / "exit.txt").read_text().strip() != "0":
                raise RuntimeError(f"Kernel {operation} failed; see {state / (operation + '.log')}")
            value = json.loads((public / "result.json").read_text())
            (state / f"{operation}.json").write_text(json.dumps(value, indent=2) + "\n")
            return value

        before = invoke("inspect")
        if args.operation == "inspect":
            report.update(before)
        else:
            # Status is a precondition; the service's owner.lock closes the start/upgrade race.
            require_stopped(status(), rollback=args.operation == "rollback")
            source = args.commit if args.operation == "install" else args.backup
            changed = invoke(args.operation, before["image_sha256"], candidate, source)
            for key in ("disk_inode", "disk_device", "disk_bytes"):
                if before[key] != changed[key]:
                    raise RuntimeError("Disk identity changed unexpectedly")
            report.update(changed)
            report["boot_verified"] = False
        report["status"] = "passed"
        success = True
    except Exception as error:
        report["error"] = str(error)
        raise
    finally:
        (state / "result.json").write_text(json.dumps(report, indent=2) + "\n")
        print(json.dumps(report, indent=2))
        if success:
            shell(f"rm -rf -- {stage}")
            shutil.rmtree(public)


if __name__ == "__main__":
    main()

File: guest/arch/kernel-upgrade/KernelUpgrade.java
import java.io.*;
import java.nio.ByteBuffer;
import java.nio.channels.*;
import java.nio.file.*;
import java.nio.file.attribute.PosixFilePermissions;
import java.security.MessageDigest;
import java.util.*;

/** Kernel-only maintenance under the same POSIX record lock as ArchVmUserService. */
public final class KernelUpgrade {
    static final LinkOption NOFOLLOW = LinkOption.NOFOLLOW_LINKS;
    static final Path BASE = Paths.get("/data/local/tmp/termux-arch-v2");

    static void validate(Path path, boolean directory, int uid) throws Exception {
        if (!path.toAbsolutePath().equals(path.toRealPath()))
            throw new IOException("Symlink or noncanonical path: " + path);
        Map<String, Object> st = Files.readAttributes(path, "unix:uid,mode,nlink", NOFOLLOW);
        if (((Number) st.get("uid")).intValue() != uid
                || (((Number) st.get("mode")).intValue() & 0077) != 0
                || (directory ? !Files.isDirectory(path, NOFOLLOW)
                : !Files.isRegularFile(path, NOFOLLOW) || ((Number) st.get("nlink")).intValue() != 1))
            throw new IOException("Unsafe owner, permissions or file type: " + path);
    }

    static String digest(Path path) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        byte[] buffer = new byte[65536];
        try (InputStream input = Files.newInputStream(path)) {
            int count;
            while ((count = input.read(buffer)) != -1) md.update(buffer, 0, count);
        }
        StringBuilder result = new StringBuilder();
        for (byte value : md.digest()) result.append(String.format(Locale.ROOT, "%02x", value & 255));
        return result.toString();
    }

    static void image(Path path, String expected, int uid) throws Exception {
        validate(path, false, uid);
        long size = Files.size(path);
        if (size < 4096 || size > 256L * 1024 * 1024)
            throw new IOException("Invalid Image size");
        try (RandomAccessFile input = new RandomAccessFile(path.toFile(), "r")) {
            input.seek(56);
            if (input.readInt() != 0x41524d64) throw new IOException("Not an ARM64 Linux Image");
        }
        if (!expected.matches("[0-9a-f]{64}") || !digest(path).equals(expected))
            throw new IOException("Image checksum mismatch");
    }

    static void syncDirectory(Path path) throws Exception {
        try (FileChannel channel = FileChannel.open(path, StandardOpenOption.READ)) { channel.force(true); }
    }

    static void copyDurable(Path source, Path destination) throws Exception {
        // CREATE_NEW: a crash can leave a partial snapshot but cannot overwrite one.
        try (InputStream input = Files.newInputStream(source);
             FileChannel out = FileChannel.open(destination, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)) {
            Files.setPosixFilePermissions(destination, PosixFilePermissions.fromString("rw-------"));
            byte[] buffer = new byte[65536];
            int count;
            while ((count = input.read(buffer)) != -1) {
                ByteBuffer bytes = ByteBuffer.wrap(buffer, 0, count);
                while (bytes.hasRemaining()) out.write(bytes);
            }
            out.force(true);
        }
    }

    static void writeDurable(Path destination, String text) throws Exception {
        try (FileChannel out = FileChannel.open(destination, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)) {
            Files.setPosixFilePermissions(destination, PosixFilePermissions.fromString("rw-------"));
            ByteBuffer bytes = ByteBuffer.wrap(text.getBytes("UTF-8"));
            while (bytes.hasRemaining()) out.write(bytes);
            out.force(true);
        }
    }

    static String operate(Path base, Path stage, String mode, String expectedCurrent,
                          String expectedCandidate, String source, int uid) throws Exception {
        validate(base, true, uid);
        validate(stage, true, uid);
        Path owner = base.resolve("owner.lock");
        // Never create or replace the service lock: both owners must lock the same inode.
        validate(owner, false, uid);
        try (RandomAccessFile file = new RandomAccessFile(owner.toFile(), "rw");
             FileLock lock = file.getChannel().tryLock()) {
            if (lock == null) throw new IOException("Arch is owned; stop it before kernel maintenance");
            Path current = base.resolve("Image");
            validate(current, false, uid);
            String before = digest(current);
            image(current, before, uid);
            // Only stat the disk. Never open, copy, truncate or replace it.
            Path disk = base.resolve("arch-rootfs.img");
            validate(disk, false, uid);
            Map<String, Object> diskStat = Files.readAttributes(disk, "unix:ino,dev,size", NOFOLLOW);
            if (mode.equals("inspect")) return report(mode, before, before, "", diskStat);
            if (!mode.equals("install") && !mode.equals("rollback")) throw new IOException("Invalid operation");
            if (!before.equals(expectedCurrent)) throw new IOException("Installed kernel changed; inspect again");
            Path backups = base.resolve("kernel-backups");
            if (!Files.exists(backups, NOFOLLOW))
                Files.createDirectory(backups, PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rwx------")));
            validate(backups, true, uid);
            Path candidate;
            String wanted;
            if (mode.equals("install")) {
                if (!source.matches("[0-9a-f]{40}")) throw new IOException("Full source commit required");
                candidate = stage.resolve("Image");
                wanted = expectedCandidate;
            } else {
                if (!source.matches("[0-9a-f]{32}")) throw new IOException("Invalid backup ID");
                Path saved = backups.resolve(source);
                validate(saved, true, uid);
                Path checksum = saved.resolve("Image.sha256");
                validate(checksum, false, uid);
                if (Files.size(checksum) != 65) throw new IOException("Invalid backup checksum");
                wanted = new String(Files.readAllBytes(checksum), "US-ASCII").trim();
                candidate = saved.resolve("Image");
            }
            image(candidate, wanted, uid);
            if (wanted.equals(before)) return report("unchanged", before, before, "", diskStat);
            String id = UUID.randomUUID().toString().replace("-", "");
            Path backup = backups.resolve(id);
            Files.createDirectory(backup, PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString("rwx------")));
            copyDurable(current, backup.resolve("Image"));
            image(backup.resolve("Image"), before, uid);
            writeDurable(backup.resolve("Image.sha256"), before + "\n");
            writeDurable(backup.resolve("transaction.txt"), "operation=" + mode + "\nsource=" + source
                    + "\nbefore=" + before + "\nafter=" + wanted + "\n");
            syncDirectory(backup);
            syncDirectory(backups);
            // Persist snapshot and its directory before making the candidate visible.
            syncDirectory(base);
            Path pending = base.resolve("Image.pending-" + id);
            copyDurable(candidate, pending);
            image(pending, wanted, uid);
            Files.move(pending, current, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
            syncDirectory(base);
            if (!digest(current).equals(wanted)) throw new IOException("Post-install verification failed");
            return report(mode, before, wanted, id, diskStat);
        }
    }

    static String report(String mode, String before, String after, String backup, Map<String, Object> disk) {
        return "{\"operation\":\"" + mode + "\",\"before_sha256\":\"" + before
                + "\",\"image_sha256\":\"" + after + "\",\"backup_id\":\"" + backup
                + "\",\"disk_inode\":" + disk.get("ino") + ",\"disk_device\":" + disk.get("dev")
                + ",\"disk_bytes\":" + disk.get("size") + "}";
    }

    public static void main(String[] args) {
        try { run(args); }
        catch (Exception error) {
            // app_process sends uncaught exceptions to logcat and may only print
            // "Killed" to the caller. Keep the actionable failure in our log.
            error.printStackTrace(System.err);
            System.err.flush();
            System.exit(1);
        }
    }

    static void run(String[] args) throws Exception {
        if (args.length != 5) throw new IllegalArgumentException("MODE STAGE CURRENT_SHA CANDIDATE_SHA SOURCE");
        int uid = (Integer) Class.forName("android.os.Process").getMethod("myUid").invoke(null);
        if (uid != 2000) throw new SecurityException("Shell Shizuku required");
        if (!args[1].matches("/data/local/tmp/aether-kernel-[0-9a-f]{32}"))
            throw new SecurityException("Invalid maintenance staging directory");
        Path stage = Paths.get(args[1]);
        String result = operate(BASE, stage, args[0], args[2], args[3], args[4], uid);
        writeDurable(stage.resolve("result.json"), result + "\n");
    }
}
